subprojects/testing-jvm/src/main/java/org/gradle/api/internal/tasks/testing/junit/JUnitTestFramework.java

```diff
@@ -42,12 +45,39 @@ public class JUnitTestFramework implements TestFramework {
     private JUnitDetector detector;
     private final DefaultTestFilter filter;
     private final boolean useImplementationDependencies;
+    private final Factory<File> testTaskTemporaryDir;
 
     public JUnitTestFramework(Test testTask, DefaultTestFilter filter, boolean useImplementationDependencies) {
+        this(filter, useImplementationDependencies, new JUnitOptions(), testTask.getTemporaryDirFactory());
+    }
+
+    private JUnitTestFramework(DefaultTestFilter filter, boolean useImplementationDependencies, JUnitOptions options, Factory<File> testTaskTemporaryDir) {
         this.filter = filter;
         this.useImplementationDependencies = useImplementationDependencies;
-        options = new JUnitOptions();
-        detector = new JUnitDetector(new ClassFileExtractionManager(testTask.getTemporaryDirFactory()));
+        this.options = options;
+        this.testTaskTemporaryDir = testTaskTemporaryDir;
+        this.detector = new JUnitDetector(new ClassFileExtractionManager(testTaskTemporaryDir));
+    }
+
+    @UsedByScanPlugin("test-retry")
+    @Override
+    public TestFramework copyWithFilters(TestFilter newTestFilters) {
+        return new JUnitTestFramework(
+            (DefaultTestFilter) newTestFilters,
+            useImplementationDependencies,
+            copyOf(options),
+            testTaskTemporaryDir
+        );
+    }
+
+    private static JUnitOptions copyOf(JUnitOptions source) {
+        // TODO: I believe we should copy the options, because they are mutable.
+        //       I don't think this would be an issue for the current use case in the test-retry
+        //       plugin, but it would be cleaner.
+        //       If you agree, I'd like to introduce a `clone` method on the TestFrameworkOptions level.
```

Comment:
I think we should do that to be safe.

Reply:
I moved the code to copy the test framework options to the respective option class.

I first planned to convert `TestFrameworkOptions` from a class to an interface an defining some kind of copy method on that level, but decided against it in order to minimize the impact.

That's why I've chosen to use a "copy constructor" way of doing things.